src/components/views/elements/MemberEventListSummary.js

```diff
@@ -52,88 +52,122 @@ module.exports = React.createClass({
-            return names + ', ' + lastName + ', and ' + remaining + ' others';
+    _renderCommaSeparatedList(items, disableAnd) {
+        if (disableAnd) {
+            return items.join(', ');
+        }
+        if (items.length === 0) {
+            return "";
+        } else if (items.length === 1) {
+            return items[0];
         } else {
-            //  name1, name2 and name3
-            return names + ' and ' + lastName;
+            let last = items.pop();
+            return items.join(', ') + ' and ' + last;
         }
     },
 
-    _renderSummary: function(joinEvents, leaveEvents) {
-        let joiners = this._renderNameList(joinEvents);
-        let leavers = this._renderNameList(leaveEvents);
+    _getDescriptionForTransition(t, plural) {
+        let beConjugated = plural ? "were" : "was";
+        let invitation = plural ? "invitations" : "an invitation";
 
-        let joinSummary = null;
-        if (joiners) {
-            joinSummary = (
-                <span>
-                    {joiners} joined the room
-                </span>
-            );
-        }
-        let leaveSummary = null;
-        if (leavers) {
-            leaveSummary = (
-                <span>
-                    {leavers} left the room
-                </span>
-            );
+        switch (t) {
+            case 'joined':  return "joined";
+            case 'left':    return "left";
+            case 'invite_reject':     return "rejected " + invitation;
+            case 'invite_withdrawal': return "withdrew " + invitation;
+            case 'invited':     return beConjugated + " invited";
+            case 'banned':      return beConjugated + " banned";
+            case 'unbanned':    return beConjugated + " unbanned";
+            case 'kicked':      return beConjugated + " kicked";
         }
 
-        // The joinEvents and leaveEvents are representative of the net movement
-        // per-user, and so it is possible that the total net movement is nil,
-        // whilst there are some events in the expanded list. If the total net
-        // movement is nil, then neither joinSummary nor leaveSummary will be
-        // truthy, so return null.
-        if (!joinSummary && !leaveSummary) {
+        return null;
+    },
+
+    _renderSummary: function(eventAggregates) {
+        let summaries = Object.keys(eventAggregates).map((transitions) => {
+            let nameList = this._renderNameList(eventAggregates[transitions]);
+
+            let repeats = 1;
+            let repeatExtra = 0;
+
+            let splitTransitions = transitions.split(',');
+            let describedTransitions = splitTransitions;
+            let plural = eventAggregates[transitions].length > 1;
+
+            for (let modulus = 1; modulus <= 2; modulus++) {
+                // Sequences that are repeating through modulus transitions will be truncated
+                if (this._isRepeatedSequence(describedTransitions, modulus)) {
```

Comment:
I don't think that looking for a pattern repeating throughout the whole of the transitions is the right approach. My understanding of matthew's intentions was that we would look specifically for pairs of 'join, leave' or 'leave, join' and treat them as special transitions. YMMV though.